### ISSUE TITLE
Investigate combinations of `unsafe` & `ref`

### ISSUE THREAD

**Issue by Nigel-Ecma** [dotnet/csharpstandard#886]
In PR #837 @KalleOlaviNiemitalo raised two comments regarding combinations of `unsafe` & `ref` that they believe should be allowed. They are being pushed to this issue to allow the PR to proceed.

> This should be allowed:
> ```
> unsafe delegate ref int D(int* p);
> ```
[link](https://github.com/dotnet/csharpstandard/pull/837#discussion_r1260615287)

> This should be allowed:
> ```
> class C {
>     unsafe ref int M(int* p) => ref *p;
> }
[link](https://github.com/dotnet/csharpstandard/pull/837#discussion_r1260615825)

---
[Associated WorkItem - 157262](https://dev.azure.com/msft-skilling/Content/_workitems/edit/157262)

**Comment by KalleOlaviNiemitalo** [comment]
This also applies to *local_function_declaration*.

**Comment by jskeet** [comment]
Meeting 2023-09-06: Assigning to Bill to investigate further.

**Comment by BillWagner** [comment]
@KalleOlaviNiemitalo 

> This also applies to _local_function_declaration_.

Can you clarify what you mean here? Are you saying we need to define rules for local functions in unsafe code, or `ref` safety in local functions?

**Comment by KalleOlaviNiemitalo** [comment]
I mean [3.6.4 Local function declarations](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/statements.md#1364-local-function-declarations) currently specifies this grammar:

```antlr
local_function_declaration
    : local_function_modifier* return_type local_function_header local_function_body
    | ref_kind ref_return_type local_function_header ref_local_function_body
    ;

local_function_header
    : identifier '(' formal_parameter_list? ')'
    | identifier type_parameter_list '(' formal_parameter_list? ')' type_parameter_constraints_clause*
    ;

local_function_modifier
    : 'async'
    | unsafe_modifier   // unsafe code support
    ;
```

This grammar does not allow using *unsafe_modifier* together with *ref_kind* on the return type.  So, for example, it does not allow the following, which should be allowed:

```csharp
class C {
     static int i;
     void M() {
         unsafe ref int L() {
             int* p = null;
             return ref i;
         }         
    }
}
```

This is only a grammar error and not related to ref-safe-context.

Could be fixed like so:

```antlr
local_function_declaration
    : local_function_modifier* return_type local_function_header local_function_body
    | ref_local_function_modifier* ref_kind ref_return_type local_function_header ref_local_function_body
    ;

local_function_header
    : identifier '(' formal_parameter_list? ')'
    | identifier type_parameter_list '(' formal_parameter_list? ')' type_parameter_constraints_clause*
    ;

local_function_modifier
    : 'async'
    | ref_local_function_modifier
    ;

ref_local_function_modifier
    : unsafe_modifier   // unsafe code support
    ;
```

Then in C# 8, 'static' could be added to *ref_local_function_modifier* and it would also become valid as *local_function_modifier*.

**Comment by KalleOlaviNiemitalo** [comment]
My two comments that were quoted in <https://github.com/dotnet/csharpstandard/issues/886#issue-1844432752> likewise apply to grammar only.  I had not even considered the ref safety aspects of pointer dereference in unsafe code.  Maybe you'll want to leave this issue open until ref safety has been reviewed, but I hope the *delegate_declaration*, *method_declaration*, and *local_function_declaration* grammar can be fixed for C# 7.3 as it seems easy and I don't think it will cause any problems in the semantics because `unsafe` can already be specified in an outer class or struct declaration.
